### ISSUE TITLE
Image Repository Allow Encrypted Images

### ISSUE THREAD

**Issue by allen-cain** [uptane/deployment-considerations#46]
The standard states that encrypted images are optional for Director repository (Section 5.3.2) & provides CONDITIONAL REQUIREMENTS for encrypted images on Director repository (Section 5.2.3.1.1).
Additionally, the standard is silent to the possibility of encrypted updates on the Image repository.
Separately, the Deployment Considerations makes reference to `"unencrypted images from the Image repository"`.

<ins>**Proposal:**</ins> Modify Deployment Considerations document to allow for both unencrypted & encrypted images on the Image repository.

**Comment by jhdalek55** [comment]
Move this back to the Standard and remove conditional text on 5.2.3.1.1. per call on 9/1/20

**Comment by jhdalek55** [comment]
@tkfu or @patrickvacek --my notes from the 9/1/20 call states that we should move this back to the Standards page and take out the conditional text. If I have this correct, can you make that move so we can do a pull request? I am assuming that these are the lines meant by the "conditional text"

"The following information is CONDITIONALLY REQUIRED for each image on the Director repository IF that image is encrypted:

Information about filenames, hashes, and file size of the encrypted image.
Information about the encryption method, and other relevant information–for example, a symmetric encryption key encrypted by the ECU’s asymmetric key could be included in the Director repository metadata."

**Comment by tkfu** [comment]
I think we aren't doing a very good job of communicating how Uptane is intended to work with encrypted images.

Let me first address the issue of encrypted images on the image repo: they just aren't very interesting, and there's not a lot of reason for Uptane to mention or care about this possibility.

Image repo targets metadata is signed with an offline key, updated infrequently, and is not customized per-vehicle, so an "encrypted" image on the image repo is definitionally going to have a decryption key that's shared across the whole fleet. Uptane doesn't really care that much about this particular use case, because you can just verify the hash of the the encrypted blob, exactly like an unencrypted image. The actual decryption of the image can be treated as part of the install step. If you want to, you can send the decryption key as custom metadata from Director, which can encrypt that key per-ECU. It doesn't really matter to any core Uptane guarantees, though; this is exactly the same as the case of, for example, UDS unlock codes or custom installation instructions.

It's when the image is encrypted dynamically, per-ECU, that we start having to be a little bit more careful. The core requirement here, for Uptane, is that there must be at least one hash that 

1. can be verified by the secondary ECU that installs the image ([Section 5.4.3.4](https://uptane.github.io/uptane-standard/uptane-standard.html#verify_image), step 7), and 
2. is included in the targets metadata from **both** the image repo and the director ([Section 5.4.4.2](https://uptane.github.io/uptane-standard/uptane-standard.html#full_verification), step 10.2.1).

Additionally, there needs to be at least one hash in the director's targets metadata that the primary can verify ([Section 5.4.2.4](https://uptane.github.io/uptane-standard/uptane-standard.html#download_images_primary)).

So in a nutshell, that's why we CONDITIONALLY REQUIRE the hash of both the original and encrypted image in director targets metadata: if it's encrypted per-ECU (which is the only "interesting" case for Uptane), the Primary isn't gonna be able to decrypt the image. But it still needs to check that the hash of the downloaded file matches something in the Uptane metadata, so the Primary checks the hash of the encrypted blob. But the primary _also_ needs the hash of the decrypted image to be in director metadata, or it can't even reach the point of downloading and decrypting the image, because then Director and Image repo wouldn't agree about what image to install. The secondary that's actually installing the image can decrypt the image, so it can check both hashes.

**Comment by tkfu** [comment]
I don't see it as an inconsistency, just a place where we aren't doing a good enough job of explaining things.

I think we could do two main things to make this better: say explicitly in the standard that we only discuss encrypted images from the Director because we only impose special requirements on images that are per-ECU encrypted, and say somewhere in the deployment considerations (and perhaps also in the standard) that there's no reason that you can't have encrypted blobs on the image repo because you treat them exactly the same as unencrypted blobs.

**Comment by jhdalek55** [comment]
Ok. If I have time later today. I'll open PRs to address what you mention above. Thanks.

Lois

**Comment by jhdalek55** [comment]
@allen-cain Can you point out in what section/page of the Deployment pages you suggest we add a mention that there is nothing to prevent the use of encrypted images on the image repository. You didn't reference a section in your original posting.

**Comment by tkfu** [comment]
Tagging this 1.1.0 for now; hopefully someone volunteers to write up the clarifications.

**Comment by jhdalek55** [comment]
@iramcdonald provided some necessary insight via email last week. Unless someone else wants to volunteer to do this, I will submit a draft PR sometime later this week adding it Deployment Best Practices in section F Enhanced Security Practices.

**Comment by jhdalek55** [comment]
Closing this Issue via Pull Request #55
